### PR TITLE
Code standards and testing updates

### PR DESCRIPTION
**Pull request by junglebarry**
Predominantly, this is a proposal to clarify our approach to testing:

* integration-tests alone are insufficient (error scenarios need unit tests)
* don't mock the system-under-test

But also:

* clarify advice on docblocks around typed languages
* some miscellaneous corrections.

### CODE REVIEW

File: topics/code-standards.md [hunk 1]
```diff
@@ -8,9 +8,8 @@ The following checklist should be applied when conducting a 90% [code review](co
 
 ### Security
 * Code should *never* contain credentials. These are a property of the environment.
-* Is the new code opening up a vulnerability? We should all be aware of methods of attack - XSS[^1] and Injection
-attacks for example.
-* Assets should be loaded with a a Protocol Relative URL[^2]
+* Is the new code opening up a vulnerability? We should all be aware of methods of attack - such as those identified by OWASP[^1].
```
Comment (junglebarry) [inline topics/code-standards.md:11]: Swapped the link to OWASP instead of enumerating by hand.

File: topics/code-standards.md [hunk 2]
```diff
@@ -102,8 +101,8 @@ of our changes on others.
 * Repetitive code - DRY [^7]
 * Large classes
 * Poor separation of concerns[^8]
-* Badly named file/class names
-* Undefined, badly named or redundant variables
+* Badly named files or classes
+* Undefined, badly named, or redundant variables
```
Comment (junglebarry) [inline topics/code-standards.md:105]: welcome to Oxford.

File: topics/code-standards.md [hunk 3]
```diff
@@ -102,8 +101,8 @@ of our changes on others.
 * Repetitive code - DRY [^7]
 * Large classes
 * Poor separation of concerns[^8]
-* Badly named file/class names
-* Undefined, badly named or redundant variables
+* Badly named files or classes
```
Comment (junglebarry) [inline topics/code-standards.md:104]: repetition

File: topics/code-standards.md [hunk 4]
```diff
@@ -27,7 +26,7 @@ you should be satisfied with the level of debug.
 * Ensure code is defensive. ```if(var[1])``` has potential to break.
 * If signatures of functions have changed, check for redundant variables. Also, does the documentation block for the function correctly reflect its intended function and the parameters it accepts and values it returns.
 * If the project supports i18n[^3] then check all user strings have been translated.[^4]
-* Look out for missing/badly named/misspelt tags.
+* Look out for missing/badly-named/misspelt tags.
```
Comment (junglebarry) [inline topics/code-standards.md:29]: I misparsed this a few times, so I've done something about it.

File: topics/code-standards.md [hunk 5]
```diff
@@ -126,19 +125,21 @@ We need to be mindful of how our applications perform on many levels. We should
...
 * Waiting for elements on a page during browser tests is much preferred to `sleep` commands.
+* Do not mock the system under test. Exceptions are only allowed when the system under test gets its external services via `getXXX` methods (but we should avoid this, preferring dependency injection).
 
 ### Documentation
-* There should be documentation - README up to date, apiary docs (if applicable), wiki page and/or infra runbook
-if necessary.
-* You should add descriptions to all functions regardless of access, define what they return and if they throw
-exceptions
-* Params should have a type and description
+* There should be living documentation - project READMEs, API documentation (if applicable), wiki pages, and/or infra runbooks. Ensure these are kept up-to-date in code changes.
```
Comment (junglebarry) [inline topics/code-standards.md:132]: Tried to ensure this is not coupled to things like apiary, but generic ideas.

File: topics/code-standards.md [hunk 6]
```diff
@@ -126,19 +125,21 @@ We need to be mindful of how our applications perform on many levels. We should
...
+* Docblocks:
+
+    * When using a typed language (e.g. Typescript), ensure your functions and methods have sensible names and appropriate parameter and return types. These are your baseline documentation. Only add docblocks if you're providing additional information, otherwise the docblock is redundant and has the potential to go stale.
+    * When using an untyped language, always include a docblock. `@param` and `@returns` must have documented types. Any tags (e.g. `@param` or `@returns` must have an accompanying description.)
+    * If a function/method can throw an error or exception, it must have a docblock with `@throws` tags for each scenario.
+
```
Comment (junglebarry) [inline topics/code-standards.md:138]: Note that our advice on docblocks is not universal, so I've updated this to reflect reality.

File: topics/testing.md [hunk 7]
```diff
@@ -11,6 +11,20 @@ For any code to be considered ready for production there need to be tests. For a
...
+
+Unit tests are a cheap way to test code, and should be included to support and document the behaviour of code. However, they are particularly useful for exercising expected error conditions (e.g. an aborted database transation; a 500 HTTP response).
+
+It can be tempting to mock the system under test, but you should not do this as it results in a restatement of the code, which is overly-coupled to code *structure* rather than *behaviour*. Mock at service boundaries instead. Furthermore, think hard before mocking at the service boundary: does the service use file, database, or network I/O? If not, could you use the real service instead?
+
+### Integration or end-to-end tests
+
+Integration and end-to-end tests are much more valuable from the perspective of testing workflows. However, they typically do not allow us to exercise specific error scenarios.
+
+Integration tests or end-to-end tests *alone * are insufficient - things that cannot be exercised in these tests **must** be covered with unit tests.
```
Comment (junglebarry) [inline topics/testing.md:27]: There's been some confusion about what's appropriate for testing - "integration tests alone are sufficient". Hopefully this should clear that up a little bit.

**Review by camillef — APPROVED** [review]

File: topics/code-standards.md [hunk 8]
```diff
@@ -8,9 +8,8 @@ The following checklist should be applied when conducting a 90% [code review](co
...
 ### Security
 * Code should *never* contain credentials. These are a property of the environment.
-* Is the new code opening up a vulnerability? We should all be aware of methods of attack - XSS[^1] and Injection
-attacks for example.
-* Assets should be loaded with a a Protocol Relative URL[^2]
+* Is the new code opening up a vulnerability? We should all be aware of methods of attack - such as those identified by OWASP[^1]. 
+* Assets should be loaded with a protocol-relative URL[^2]
 
 ### General
 * Fix any broken windows. They are not someone else's problem. See [charter](charter.html).
```
Comment (foomatty) [inline topics/code-standards.md:15]: Out of scope, but I think we need to update the charter re broken windows.

**Review by foomatty — APPROVED** [review]

File: topics/code-standards.md [hunk 9]
```diff
@@ -8,9 +8,8 @@ The following checklist should be applied when conducting a 90% [code review](co
 
 ### Security
 * Code should *never* contain credentials. These are a property of the environment.
-* Is the new code opening up a vulnerability? We should all be aware of methods of attack - XSS[^1] and Injection
-attacks for example.
-* Assets should be loaded with a a Protocol Relative URL[^2]
+* Is the new code opening up a vulnerability? We should all be aware of methods of attack - such as those identified by OWASP[^1]. 
+* Assets should be loaded with a protocol-relative URL[^2]
```
Comment (danielmatthew) [inline topics/code-standards.md:12]: I understood this to be bad practice, and [indeed the link now declares it to be an anti-pattern](https://www.paulirish.com/2010/the-protocol-relative-url/).

> Now that SSL is encouraged for everyone and doesn’t have performance concerns, this technique is now an anti-pattern. If the asset you need is available on SSL, then always use the https:// asset.

File: topics/code-standards.md (hunk 3 above)
Comment (danielmatthew) [reply topics/code-standards.md:104]: You updated "badly named" to "badly-named" on L29.

File: topics/code-standards.md (hunk 4 above)
Comment (danielmatthew) [reply topics/code-standards.md:29]:
```suggestion
* Look out for missing, badly-named, or misspelt tags.
```

File: topics/code-standards.md (hunk 2 above)
Comment (danielmatthew) [reply topics/code-standards.md:105]: "badly named" vs "badly-named"

File: topics/code-standards.md (hunk 3 above)
Comment (junglebarry) [reply topics/code-standards.md:104]: I like the cut of your jib! Will amend.

File: topics/code-standards.md (hunk 9 above)
Comment (junglebarry) [reply topics/code-standards.md:12]: Heh. I’d actually misread this to be a relative URL, not a protocol-relative URL. Relative URLs make a lot of sense in many our our APIs. Protocol relative is a bare minimum; HTTPS is preferred. I’ll amend.

**Pull request merged**